### ISSUE TITLE
Segmentation result quality is subpar

### ISSUE THREAD

**Issue by nylocx** [floe/backscrub#29]
Hi, first thanks for the project and trying to bring background removal to linux, but for me the result is really bad and I don't know why.
For me it often looks like that:
![image](https://user-images.githubusercontent.com/3329124/104724751-64bc2d00-5731-11eb-8284-c1f5b1f6a0fa.png)

Here is the debug output of the deepseg binary:
```
./deepseg -d -d -c /dev/video0 -v /dev/video4
deepseg v0.2.0
(c) 2021 by floe@butterbrot.org
https://github.com/floe/deepbacksub
debug:  2
ccam:   /dev/video0
vcam:   /dev/video4
width:  640
height: 480
back:   images/background.png
threads:2
model:  models/segm_full_v679.tflite

vid_format->type                = 2
vid_format->fmt.pix.width       = 640
vid_format->fmt.pix.height      = 480
vid_format->fmt.pix.pixelformat = 1448695129
vid_format->fmt.pix.sizeimage   = 614400
vid_format->fmt.pix.field       = 1
vid_format->fmt.pix.bytesperline= 1280
vid_format->fmt.pix.colorspace  = 8

tensor #0: 1
tensor #0: 144
tensor #0: 256
tensor #0: 3
tensor #244: 1
tensor #244: 144
tensor #244: 256
tensor #244: 2
```

And this is from v4l2-ctl --all -d /dev/video0
```
Driver name      : uvcvideo
Video input : 0 (Camera 1: ok)
Format Video Capture:
        Width/Height      : 640/480
        Pixel Format      : 'YUYV' (YUYV 4:2:2)
        Field             : None
        Bytes per Line    : 1280
        Size Image        : 614400
        Colorspace        : sRGB
        Transfer Function : Rec. 709
        YCbCr/HSV Encoding: ITU-R 601
        Quantization      : Default (maps to Limited Range)
        Flags             : 
```
```
opencv_version
4.5.1
```
So looks ok for me

Any ideas where I should start debugging this?

**Comment by floe** [comment]
Two things to try: 

- improve overall lighting for the scene (I've had similar issues with the Google Meet network)
- try another network from models/ (e.g. `-m models/deeplabv3_257_mv_gpu.tflite` - will be slower, though)

**Comment by shannonlowder** [comment]
I have a similar problem.  I seem to get the best results when using ./models/deeplabv3_257_mv_gpu.tflite, running at 640x480. I've tried throwing more threads at the problem and the process runs at 7-9 fps.  I've tried 2, 4, 8, 16, 32, 64, 128, and 256 threads. 16 threads seem to be where the performance taps out.   

I have plenty of CPU available on this box. (48 cores plus hyperthreading).

**Comment by floe** [comment]
Out of curiosity, did you try the body-pix model?

**Comment by shannonlowder** [comment]
I did.  [https://1drv.ms/u/s!AgvTlkbvPHhyk5g-6XJxEaJJKUlmnA](https://1drv.ms/u/s!AgvTlkbvPHhyk5g-6XJxEaJJKUlmnA)

The result loses several "blocks" of my face.

**Comment by floe** [comment]
AFAICT the Google Meet network was trained with "selfies", i.e. pictures from phone front cameras. Those have a very wide FOV and are portrait format. I've tried to approximate this by selecting a portrait-format ROI from the landscape-format webcam image, but that may not be enough. In your example image, your portrait is filling most of the picture width. Does the `segm_xxx` default model work better if you move away a bit?

**Comment by shannonlowder** [comment]
I had my webcam set to zoom in a little for conference calls. I  zoomed out and stood back from the camera a bit.  The segmentation seems worse on both of the segm_xxx models.  I'm continuing to play with the different models and different settings to get a clearer result.

**Comment by BenBE** [comment]
On a different note: I noticed that depending on the hue/saturation and overall lighting of the image the detection even for the same net can range from "near perfect" to "not at all" …If you have the chance, try to calibrate the hue/saturation settings of your cam while running deepbacksub to check the result. In my case, configuring the webcame device in OBS Studio, while watchung deepbacksub process its image gave a quite good feel for which settings worked best.

**Comment by shannonlowder** [comment]
I was able to get a passable result by playing with the settings as you've suggested!  I'll keep playing with the settings to see how much further I can improve!

**Comment by mikaelhg** [comment]
The results have been pretty poor for me so far, but the best settings I've found to date have been:

```
$ v4l2-ctl -D
Driver Info:
	Driver name      : uvcvideo
	Card type        : Microsoft® LifeCam HD-3000: Mi
	Bus info         : usb-0000:00:14.0-5
	Driver version   : 5.8.18
	Capabilities     : 0x84a00001
		Video Capture
		Metadata Capture
		Streaming
		Extended Pix Format
		Device Capabilities
	Device Caps      : 0x04200001
		Video Capture
		Streaming
		Extended Pix Format

$ v4l2-ctl --list-ctrls
                     brightness 0x00980900 (int)    : min=30 max=255 step=1 default=133 value=150
                       contrast 0x00980901 (int)    : min=0 max=10 step=1 default=5 value=10
                     saturation 0x00980902 (int)    : min=0 max=200 step=1 default=83 value=100
 white_balance_temperature_auto 0x0098090c (bool)   : default=1 value=1
           power_line_frequency 0x00980918 (menu)   : min=0 max=2 default=2 value=0
      white_balance_temperature 0x0098091a (int)    : min=2800 max=10000 step=1 default=4500 value=4500 flags=inactive
                      sharpness 0x0098091b (int)    : min=0 max=50 step=1 default=25 value=25
         backlight_compensation 0x0098091c (int)    : min=0 max=10 step=1 default=0 value=0
                  exposure_auto 0x009a0901 (menu)   : min=0 max=3 default=1 value=3
              exposure_absolute 0x009a0902 (int)    : min=5 max=20000 step=1 default=156 value=156 flags=inactive
                   pan_absolute 0x009a0908 (int)    : min=-201600 max=201600 step=3600 default=0 value=0
                  tilt_absolute 0x009a0909 (int)    : min=-201600 max=201600 step=3600 default=0 value=0
                  zoom_absolute 0x009a090d (int)    : min=0 max=10 step=1 default=0 value=0


```

**Comment by mikaelhg** [comment]
The results seem to be extremely, extremely dependent on the specific color spaces, gradients, camera sensors, lighting, etc.

**Comment by floe** [comment]
Anyone who's got a bit of experience with reverse-engineering Javascript code might try to have a look at what the Google Meet web app does with the camera picture before passing it on to the TFlite model. Unfortunately, this is not at all my area...

**Comment by floe** [comment]
Also, 5b4eacfaf93eedffe04634e616e6c1f24bc0a659 might help to improve this a bit, at least in cases where the segmentation works okay-ish but flickers.

**Comment by mikaelhg** [comment]
How about building a test harness where we could submit video clips that are failing for us, so that you can actually reproduce the issues locally?

**Comment by floe** [comment]
Not a bad idea. You can actually attach files to Github issues by now, not sure if there is a size limit - maybe just give it a try?

**Comment by BenBE** [comment]
> Also, [5b4eacf](https://github.com/floe/deepbacksub/commit/5b4eacfaf93eedffe04634e616e6c1f24bc0a659) might help to improve this a bit, at least in cases where the segmentation works okay-ish but flickers.

With the WMA the detection is much more stable with the downside of a slight "motion lag": When you move your head to one side you can notice a slight delay until the mask has caught up with the movement, resulting in a small part of the side of the face cut off. With slow movements this is barely noticeable, but with fast movements this is quite visible.

Also with the WMA applied some minor "hole" stay within the masked image. the easiest way to reproduce this for me was to angel one arm upward, so the cam picks up both my face, upper body and the full arm+hand.
![image](https://user-images.githubusercontent.com/111196/106304437-a255a580-625b-11eb-8e44-926da9ae4ad2.png)
deepbacksub currently seems to start dropping the arm at some point.
(Image not of me, just to demonstrate …)

**Comment by floe** [comment]
I guess at some point we're hitting the limitations of the segmentation models. If you look at the description of the GoogleMeet model (https://ai.googleblog.com/2020/10/background-features-in-google-meet.html), they use various image-space tricks (soft edges etc.) to make it look less obvious if the segmentation falls short (e.g. I'm pretty sure that the animations for the blur don't capture the raised hands either, it's just not quite as noticeable).

It would be very much worth retraining these models with more diverse open data, but that's a pretty big project on its own...

**Comment by phlash** [comment]
@nylocx, @shannonlowder, @mikaelhg (and anyone else experiencing iffy segmentation): have you tried the latest `selfiesegmentation_mlkit-256x256-2021_01_19-v1215.f16.tflite` model? I find this works much better for my quite poor camera (Lenovo ThinkPad T430 built-in). We are also experimenting with pre and post processing the video/mask streams in the `experimental` branch.

**Comment by mikaelhg** [comment]
Sorry it took a while, @phlash, but I finally had time to test it, and you're right, it's much MUCH better with the new model. There's slight headrest flicker, but otherwise this is now usable.

**Comment by mikaelhg** [comment]
Maybe the model could be finetuned with an additional dataset containing images with chair headrests and other office equipment.

**Comment by floe** [comment]
@mikaelhg in principle, yes - although you certainly need a beefy NVidia GPU to even begin with retraining, and you also need to label the data manually up front, so this is a pretty serious effort...

**Comment by mikaelhg** [comment]
Anyway, IMO with the new model, the symptoms which led to the opening of this issue have been alleviated. I'll create a new issue for the model finetuning.

**Comment by floe** [comment]
Thanks - closing for now. We'll probably set the selfiesegmentation model as new default.
